src/ext/circuit_breaker.c

```diff
@@ -0,0 +1,137 @@
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
+dd_trace_circuit_breaker_t local_dd_trace_circuit_breaker = {{0}};
+#pragma GCC diagnostic pop
+#else  //__clang__
+dd_trace_circuit_breaker_t local_dd_trace_circuit_breaker = {0};
+#endif
+
+static void handle_perpare_error(const char *call_name) {
+    perror(call_name);
+
+    if (!dd_trace_circuit_breaker) {
+        // if shared memory is not working use local copy
+        dd_trace_circuit_breaker = &local_dd_trace_circuit_breaker;
+    }
+}
+
+static void prepare_cb() {
+    if (!dd_trace_circuit_breaker) {
+        int shm_fd = shm_open(DD_TRACE_CIRCUIT_BREAKER_SHMEM_KEY, O_CREAT | O_RDWR, 0666);
+        if (shm_fd < 0) {
+            handle_perpare_error("shm_open");
+            return;
+        }
+
+        // initialize size if not yet initialized
+        // fetch stat with shmem size
+        struct stat stats;
+        if (fstat(shm_fd, &stats) != 0) {
+            handle_perpare_error("fstat");
+            return;
+        }
+        // do the resize if size is too small
+        if (stats.st_size < 0 || (size_t)stats.st_size < sizeof(dd_trace_circuit_breaker_t)) {
+            if (ftruncate(shm_fd, sizeof(dd_trace_circuit_breaker_t)) != 0) {
+                handle_perpare_error("ftruncate");
+                return;
+            }
+        }
+        // mmap shared memory to local memory
+        dd_trace_circuit_breaker_t *shared_breaker =
+            mmap(NULL, sizeof(dd_trace_circuit_breaker_t), PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
+        if (shared_breaker == MAP_FAILED) {
+            handle_perpare_error("mmap");
+
+            return;
+        }
+
+        dd_trace_circuit_breaker = shared_breaker;
+    }
+}
+
+static uint64_t current_timestamp_monotonic_usec() {
+    struct timespec t;
+    clock_gettime(CLOCK_MONOTONIC, &t);
+
+    return t.tv_sec * 1000 * 1000 + t.tv_nsec / 1000;
+}
+
+static int64_t get_max_consecutive_failures() {
+    return ddtrace_get_int_config(DD_TRACE_CIRCUIT_BREAKER_ENV_MAX_CONSECUTIVE_FAILURES,
+                                  DD_TRACE_CIRCUIT_BREAKER_DEFAULT_MAX_CONSECUTIVE_FAILURES);
+}
+
+static int64_t get_retry_time_usec() {
+    return ddtrace_get_int_config(DD_TRACE_CIRCUIT_BREAKER_ENV_RETRY_TIME_MSEC,
+                                  DD_TRACE_CIRCUIT_BREAKER_DEFAULT_RETRY_TIME_MSEC) *
+           1000;
+}
+
+uint32_t dd_tracer_circuit_breaker_can_try() {
+    if (dd_tracer_circuit_breaker_is_closed()) {
+        return 1;
+    }
+    uint64_t last_failure_timestamp = atomic_load(&dd_trace_circuit_breaker->last_failure_timestamp);
+    uint64_t current_time = current_timestamp_monotonic_usec();
+
+    return (last_failure_timestamp + get_retry_time_usec()) <= current_time;
+}
+
+void dd_tracer_circuit_breaker_register_error() {
+    prepare_cb();
+
+    atomic_fetch_add(&dd_trace_circuit_breaker->consecutive_failures, 1);
+    atomic_fetch_add(&dd_trace_circuit_breaker->total_failures, 1);
```

Comment:
I left in in since this is the data I would love to collect with runtime metrics. But right now it doesn't serve any purpose.

However I think actually might be good idea to add debug mode, where this data could be exposed.

Reply:
I ❤️ the idea of putting that at debug level. :)